File: C++/maximum-candies-you-can-get-from-boxes.cpp
// Time:  O(n^2)
// Space: O(n)

class Solution {
public:
    int maxCandies(vector<int>& status, vector<int>& candies, vector<vector<int>>& keys, vector<vector<int>>& containedBoxes, vector<int>& initialBoxes) {
        int result = 0;
        queue<int> q;
        for (const auto& box : initialBoxes) {
            q.emplace(box);
        }
        while (!q.empty()) {
            bool changed = false;
            for (int i = q.size() - 1; i >= 0; --i) {
                const auto box = q.front(); q.pop();
                if (!status[box]) {
                    q.emplace(box);
                    continue;
                }
                changed = true;
                result += candies[box];
                for (const auto& contained_key : keys[box]) {
                    status[contained_key] = 1;
                }
                for (const auto& contained_box : containedBoxes[box]) {
                    q.emplace(contained_box);
                }
            }
            if (!changed) {
                break;
            }
        }
        return result;
    }
};

File: Python/maximum-candies-you-can-get-from-boxes.py
# Time:  O(n^2)
# Space: O(n)

import collections


class Solution(object):
    def maxCandies(self, status, candies, keys, containedBoxes, initialBoxes):
        """
        :type status: List[int]
        :type candies: List[int]
        :type keys: List[List[int]]
        :type containedBoxes: List[List[int]]
        :type initialBoxes: List[int]
        :rtype: int
        """
        result = 0
        q = collections.deque(initialBoxes)
        while q:
            changed = False
            for _ in xrange(len(q)):
                box = q.popleft()
                if not status[box]:
                    q.append(box)
                    continue
                changed = True
                result += candies[box]
                for contained_key in keys[box]:
                    status[contained_key] = 1
                for contained_box in containedBoxes[box]:
                    q.append(contained_box)
            if not changed:
                break
        return result
